File: articles/perseverance-of-effort-versus-consistency-of-interests/python/grit_facet_comparison_model.py
"""
Perseverance of Effort Versus Consistency of Interests

Synthetic-data workflow for modeling grit facets separately, comparing total
grit with facet-level models, and interpreting outcomes with context.

This script is for article support and research-method demonstration only.
It should not be used to assess, rank, hire, discipline, or evaluate real people.
"""

from pathlib import Path

import numpy as np
import pandas as pd
import statsmodels.api as sm


ROOT = Path(__file__).resolve().parents[1]
DATA_OUT = ROOT / "data" / "processed" / "grit-facet-comparison-modeled-data-python.csv"
TABLE_OUT = ROOT / "outputs" / "tables" / "grit-facet-comparison-models-python.csv"
PROFILE_OUT = ROOT / "outputs" / "tables" / "grit-facet-profile-summary-python.csv"

rng = np.random.default_rng(42)
n = 900

perseverance_effort = rng.normal(0, 1, n)
consistency_interests = rng.normal(0, 1, n)

conscientiousness = 0.62 * perseverance_effort + rng.normal(0, 0.80, n)
self_control = 0.45 * perseverance_effort + rng.normal(0, 0.90, n)
social_support = rng.normal(0, 1, n)
prior_achievement = rng.normal(0, 1, n)
burnout = rng.normal(0, 1, n)

grit_total = 0.60 * perseverance_effort + 0.40 * consistency_interests

long_term_outcome = (
    0.28 * perseverance_effort
    + 0.08 * consistency_interests
    + 0.25 * prior_achievement
    + 0.18 * conscientiousness
    + 0.20 * social_support
    - 0.22 * burnout
    + rng.normal(0, 1, n)
)

df = pd.DataFrame(
    {
        "perseverance_effort": perseverance_effort,
        "consistency_interests": consistency_interests,
        "grit_total": grit_total,
        "conscientiousness": conscientiousness,
        "self_control": self_control,
        "social_support": social_support,
        "prior_achievement": prior_achievement,
        "burnout": burnout,
        "long_term_outcome": long_term_outcome,
    }
)

pe_median = df["perseverance_effort"].median()
ci_median = df["consistency_interests"].median()

conditions = [
    (df["perseverance_effort"] >= pe_median) & (df["consistency_interests"] >= ci_median),
    (df["perseverance_effort"] >= pe_median) & (df["consistency_interests"] < ci_median),
    (df["perseverance_effort"] < pe_median) & (df["consistency_interests"] >= ci_median),
    (df["perseverance_effort"] < pe_median) & (df["consistency_interests"] < ci_median),
]
choices = [
    "high_effort_high_consistency",
    "high_effort_low_consistency",
    "low_effort_high_consistency",
    "low_effort_low_consistency",
]
df["facet_profile"] = np.select(conditions, choices, default="unclassified")

DATA_OUT.parent.mkdir(parents=True, exist_ok=True)
TABLE_OUT.parent.mkdir(parents=True, exist_ok=True)
PROFILE_OUT.parent.mkdir(parents=True, exist_ok=True)
df.to_csv(DATA_OUT, index=False)

model_total_grit = sm.OLS(
    df["long_term_outcome"],
    sm.add_constant(df[["grit_total"]]),
).fit()

model_facets = sm.OLS(
    df["long_term_outcome"],
    sm.add_constant(df[["perseverance_effort", "consistency_interests"]]),
).fit()

model_contextual = sm.OLS(
    df["long_term_outcome"],
    sm.add_constant(
        df[
            [
                "perseverance_effort",
                "consistency_interests",
                "prior_achievement",
                "conscientiousness",
                "self_control",
                "social_support",
                "burnout",
            ]
        ]
    ),
).fit()

comparison = pd.DataFrame(
    {
        "model": ["total_grit_only", "facets_only", "facets_plus_context"],
        "r_squared": [
            model_total_grit.rsquared,
            model_facets.rsquared,
            model_contextual.rsquared,
        ],
        "adjusted_r_squared": [
            model_total_grit.rsquared_adj,
            model_facets.rsquared_adj,
            model_contextual.rsquared_adj,
        ],
    }
)

profile_summary = (
    df.groupby("facet_profile", as_index=False)
    .agg(
        n=("facet_profile", "size"),
        mean_perseverance_effort=("perseverance_effort", "mean"),
        mean_consistency_interests=("consistency_interests", "mean"),
        mean_long_term_outcome=("long_term_outcome", "mean"),
        mean_burnout=("burnout", "mean"),
        mean_social_support=("social_support", "mean"),
    )
)

comparison.to_csv(TABLE_OUT, index=False)
profile_summary.to_csv(PROFILE_OUT, index=False)

print("Saved modeled dataset:", DATA_OUT)
print("Saved model comparison:", TABLE_OUT)
print("Saved profile summary:", PROFILE_OUT)
print(comparison.round(4))
print(profile_summary.round(4))

File: articles/perseverance-of-effort-versus-consistency-of-interests/r/grit-facet-comparison-model.R
# Perseverance of Effort Versus Consistency of Interests
# Synthetic-data workflow for research-method demonstration only.
# Do not use this workflow to evaluate, rank, hire, discipline, or assess real people.

set.seed(42)

n <- 900

perseverance_effort <- rnorm(n)
consistency_interests <- rnorm(n)

conscientiousness <- 0.62 * perseverance_effort + rnorm(n, sd = 0.80)
self_control <- 0.45 * perseverance_effort + rnorm(n, sd = 0.90)
social_support <- rnorm(n)
prior_achievement <- rnorm(n)
burnout <- rnorm(n)

grit_total <- 0.60 * perseverance_effort + 0.40 * consistency_interests

long_term_outcome <- (
  0.28 * perseverance_effort +
  0.08 * consistency_interests +
  0.25 * prior_achievement +
  0.18 * conscientiousness +
  0.20 * social_support -
  0.22 * burnout +
  rnorm(n)
)

df <- data.frame(
  perseverance_effort,
  consistency_interests,
  grit_total,
  conscientiousness,
  self_control,
  social_support,
  prior_achievement,
  burnout,
  long_term_outcome
)

pe_median <- median(df$perseverance_effort)
ci_median <- median(df$consistency_interests)

df$facet_profile <- ifelse(
  df$perseverance_effort >= pe_median & df$consistency_interests >= ci_median,
  "high_effort_high_consistency",
  ifelse(
    df$perseverance_effort >= pe_median & df$consistency_interests < ci_median,
    "high_effort_low_consistency",
    ifelse(
      df$perseverance_effort < pe_median & df$consistency_interests >= ci_median,
      "low_effort_high_consistency",
      "low_effort_low_consistency"
    )
  )
)

dir.create("data/processed", recursive = TRUE, showWarnings = FALSE)
dir.create("outputs/tables", recursive = TRUE, showWarnings = FALSE)

write.csv(df, "data/processed/grit-facet-comparison-modeled-data-r.csv", row.names = FALSE)

profile_summary <- aggregate(
  cbind(long_term_outcome, perseverance_effort, consistency_interests, burnout, social_support) ~ facet_profile,
  data = df,
  FUN = mean
)

write.csv(profile_summary, "outputs/tables/grit-facet-profile-summary-r.csv", row.names = FALSE)

model_total_grit <- lm(long_term_outcome ~ grit_total, data = df)

model_facets <- lm(
  long_term_outcome ~ perseverance_effort + consistency_interests,
  data = df
)

model_contextual <- lm(
  long_term_outcome ~ perseverance_effort + consistency_interests +
    prior_achievement + conscientiousness + self_control +
    social_support + burnout,
  data = df
)

comparison <- data.frame(
  model = c("total_grit_only", "facets_only", "facets_plus_context"),
  r_squared = c(
    summary(model_total_grit)$r.squared,
    summary(model_facets)$r.squared,
    summary(model_contextual)$r.squared
  ),
  adjusted_r_squared = c(
    summary(model_total_grit)$adj.r.squared,
    summary(model_facets)$adj.r.squared,
    summary(model_contextual)$adj.r.squared
  )
)

write.csv(comparison, "outputs/tables/grit-facet-comparison-models-r.csv", row.names = FALSE)

print(round(profile_summary, 4))
print(round(comparison, 4))
